controller/api/destination/hostport_adaptor.go

```diff
@@ -0,0 +1,257 @@
+		k8sAPI:              k8sAPI,
+		metadataAPI:         metadataAPI,
+		metrics:             metrics,
+		log:                 log,
+	}
+}
+
+func (pt *hostPortAdaptor) Sync() error {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	opaquePorts := getAnnotatedOpaquePorts(pt.address.Pod, pt.defaultOpaquePorts)
+	endpoint, err := pt.createEndpoint(*pt.address, opaquePorts)
+	if err != nil {
+		return fmt.Errorf("failed to create endpoint: %w", err)
+	}
+	pt.endpoint = endpoint
+	pt.log.Debugf("Sync for endpoint %s", pt.endpoint)
+	pt.subscribed = false
+
+	// If the endpoint's port is annotated as opaque, we don't need to
+	// subscribe for updates because it will always be opaque
+	// regardless of any Servers that may select it.
+	if _, ok := opaquePorts[pt.port]; ok {
+		pt.UpdateProtocol(true)
+	} else if pt.address.Pod == nil {
+		pt.UpdateProtocol(false)
+	} else {
+		pt.UpdateProtocol(pt.address.OpaqueProtocol)
+		pt.servers.Subscribe(pt.address.Pod, pt.port, pt)
+		pt.subscribed = true
+	}
+
+	return nil
+}
+
+func (pt *hostPortAdaptor) Clean() {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	if pt.subscribed {
+		pt.log.Debugf("Clean for endpoint %s", pt.endpoint)
+		pt.servers.Unsubscribe(pt.address.Pod, pt.port, pt)
+		pt.subscribed = false
+	}
+}
+
+func (pt *hostPortAdaptor) UpdateProtocol(opaqueProtocol bool) {
+	pt.listener.UpdateProtocol(pt.address.Pod, pt.endpoint, opaqueProtocol)
+}
+
+// Update is an informer event handler - All operations should be non-blocking
+func (pt *hostPortAdaptor) Update(pod *corev1.Pod) {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	if !pt.matchesIPPort(pod) {
+		return
+	}
+
+	if pt.podReady && pt.address.Pod.UID != pod.UID {
+		pt.log.Tracef("Current pod still ready, ignoring event on %s.%s", pod.Name, pod.Namespace)
+		return
+	}
+
+	if pt.podReady && !isRunningAndReady(pod) {
+		pt.log.Debugf("Pod %s.%s became unready - remove", pod.Name, pod.Namespace)
+		pt.updateAddress(nil)
+		return
+	}
+
+	if !pt.podReady && !isRunningAndReady(pod) {
+		pt.log.Tracef("Ignore event on %s.%s until it becomes ready", pod.Name, pod.Namespace)
+		return
+	}
+
+	if !pt.podReady && isRunningAndReady(pod) {
+		pt.log.Debugf("Pod %s.%s became ready", pod.Name, pod.Namespace)
+		pt.updateAddress(pod)
+		return
+	}
+
+	pt.log.Tracef("Ignored event on pod %s.%s", pod.Name, pod.Namespace)
+}
+
+// Remove is an informer event handler - All operations should be non-blocking
+func (pt *hostPortAdaptor) Remove(pod *corev1.Pod) {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	if !pt.matchesIPPort(pod) {
+		return
+	}
+
+	if pt.address.Pod == nil {
+		pt.log.Tracef("Pod %s.%s already removed; discard event", pod.Name, pod.Namespace)
+		return
+	}
+
+	pt.log.Debugf("Remove pod %s.%s", pod.Name, pod.Namespace)
+	pt.updateAddress(nil)
+}
+
+func (pt *hostPortAdaptor) MetricsInc() {
+	pt.metrics.Inc()
+}
+
+func (pt *hostPortAdaptor) MetricsDec() {
+	pt.metrics.Dec()
+}
+
+func (pt *hostPortAdaptor) updateAddress(pod *corev1.Pod) {
+	go func() {
```

Comment:
Should we leave concurrency to the caller here? I feel like a go routine here might make it more difficult to troubleshoot.

Reply:
It seemed more succinct to me to do it in the common method. Can you clarify how the alternative would help?

Reply:
Ok I ended up moving the goroutines creation to the consumers in my latest refactoring, like you suggested. Those functions explicitly ask for non-blocking code in their godocs, so it's more explicit to handle that right in there.

Reply:
Yep that makes sense, thanks for the clarification!